File: BeeCrowd/InfixaPosfixa.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

typedef struct pilha{
    char* array;
    int topo;
    int tam;
} Pilha;

//Método para inicializar a pilha
void inicializarPilha(Pilha *p, int tamanho){
    p->array = (char*)malloc(tamanho * sizeof(char));
    p->topo = -1;
    p->tam = tamanho;
}

//Método para destruir a pilha e liberar memória
void destruirPilha(Pilha *p){
    free(p->array);
    p->topo = -1;
    p->tam = 0;
}

//Método para inserir na pilha 
void empilhar(char caracter, Pilha *p){
    if(p->topo >= p->tam - 1){
         printf("Erro. Pilha cheia.");
         return;
    }

    p->array[++p->topo] = caracter;
}

char mostrarTopo(Pilha *p){
    if(p->topo == -1){
        return '\0';
    }

    return p->array[p->topo];
}

//Método para remover na pilha 
char desempilhar(Pilha *p){
    if(p->topo == -1){
        printf("Erro. Pilha vazia.");
        return '\0';
    }

    return p->array[p->topo--];
}

//Método para verificar a ordem de precedência do caracter
int precedencia(char caracter){
    if(caracter == '^'){
        return 1;
    } else if(caracter == '*' || caracter == '/'){
        return 2;
    } else if(caracter == '+' || caracter == '-'){
        return 3;
    } else {
        return 4;
    }
}

void posfixar(char infixa[], Pilha *p){
    char saida[300];
    int contador = 0;

    for(int i = 0; i < strlen(infixa); i++){
        char caracterAtual = infixa[i];

        // Primeiro caso: o caracter é alfanúmerico, ele deve ser adicionado a saída
        if(caracterAtual >= 'A' && caracterAtual <= 'Z' ||
         caracterAtual >= 'a' && caracterAtual <= 'z' ||
         caracterAtual >= '0' && caracterAtual <= '9'){
            saida[contador++] = caracterAtual;
        }

        //Segundo caso: verificar se o caracter é "("
        if(caracterAtual == '('){
            empilhar(caracterAtual, p);
        } 

        //Terceiro caso: verificar se o caracter é ")". Caso verdadeiro, desempilhar tudo da pilha 
        else if(caracterAtual == ')'){
            while(p->topo != -1 && mostrarTopo(p) != '('){ //Desempilhar até encontrar (
                saida[contador++] = desempilhar(p); 
            }
            if(mostrarTopo(p) == '('){
                desempilhar(p);
            }
        }

        //Quarto caso: verificar os operadores e a ordem de precedência
        else if(caracterAtual == '^' || caracterAtual == '*' || caracterAtual == '/' || caracterAtual == '+' || caracterAtual == '-'){
            while(p->topo != -1 && mostrarTopo(p) != '(' && precedencia(mostrarTopo(p)) <= precedencia(caracterAtual)){
                saida[contador++] = desempilhar(p);
            }
            empilhar(caracterAtual, p);
        }

    }

    //Por fim, imprimir os caracteres restantes
    while(p->topo != -1){
        saida[contador++] = desempilhar(p);
    }

    saida[contador] = '\0';
    printf("%s\n", saida);

}

int main(){
    int quantidade;
    scanf("%d", &quantidade);
    getchar();

    Pilha p;
    inicializarPilha(&p, 300);

    for(int i = 0; i < quantidade; i++){
        char entrada[300];
        fgets(entrada, sizeof(entrada), stdin);

        posfixar(entrada, &p);
    }
    
    destruirPilha(&p);

    return 0;
}

File: BeeCrowd/InfixaPosfixa.java
import java.util.Scanner;

class Pilha {
    char array[];
    int topo;

    public Pilha(int tam) {
        array = new char[tam];
        topo = -1;
    }

    public void empilhar(char caracter) {
        if (topo >= array.length - 1) {
            throw new IllegalStateException("Array está cheio, não é possível inserir mais elementos.");
        }

        array[++topo] = caracter;
    }

    public char desempilhar() {
        if (topo == -1) {
            throw new IllegalStateException("Array está vazio.");
        }

        return array[topo--];
    }

    public char verTopo() {
        if (topo == -1) {
            return '\0'; // Retorna caractere nulo se a pilha estiver vazia
        }

        return array[topo];
    }
}

public class InfixaPosfixa {

    public static String posfixar(String infixa) {
        Pilha pilha = new Pilha(infixa.length());
        StringBuilder str = new StringBuilder();

        for (int i = 0; i < infixa.length(); i++) {
            char caracterAtual = infixa.charAt(i);

            // Primeiro caso: o caracter é alfanúmerico, ele deve ser adicionado a saída
            if (caracterAtual >= '0' && caracterAtual <= '9' ||
                    caracterAtual >= 'A' && caracterAtual <= 'Z' ||
                    caracterAtual >= 'a' && caracterAtual <= 'z') {
                str.append(caracterAtual);
            }
            
            // Segundo caso: verificar se o caracter é "("
            else if (caracterAtual == '(') {
                pilha.empilhar(caracterAtual);
            }

            // Terceiro caso: verificar se o caracter é ")"
            else if (caracterAtual == ')') {
                while (pilha.verTopo() != '(' && pilha.topo != -1) {
                    str.append(pilha.desempilhar());
                }
                if (pilha.verTopo() == '(') {
                    pilha.desempilhar(); // Remove o '(' da pilha sem adicionar à saída
                }
            }

            // Quarto caso: verificar os operadores e a ordem de precedência
            else if (caracterAtual == '^' || caracterAtual == '*' || caracterAtual == '/' || caracterAtual == '+' || caracterAtual == '-') {
                while (pilha.topo != -1 && pilha.verTopo() != '(' && precedencia(pilha.verTopo()) <= precedencia(caracterAtual)) {
                    str.append(pilha.desempilhar());
                }
                pilha.empilhar(caracterAtual);
            }
        }

        // Adicionar os elementos restantes
        while (pilha.topo != -1) {
            str.append(pilha.desempilhar());
        }

        return str.toString();
    }

    public static int precedencia(char operador) {
        if (operador == '^') {
            return 1;
        } else if (operador == '*' || operador == '/') {
            return 2;
        } else if (operador == '+' || operador == '-') {
            return 3;
        } else {
            return 4; // Para '(' e outros caracteres
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int quantidade = sc.nextInt();
        sc.nextLine();

        for (int i = 0; i < quantidade; i++) {
            String entrada = sc.nextLine();
            System.out.println(posfixar(entrada));
        }

        sc.close();
    }
}